### PR TITLE
Fixes #24271 - Edit Host as non-admin is slower than admin

### PR DESCRIPTION
**Pull request by kgaikwad**
For non-admin users, it takes more time in processing
common_parameters partials as comparative to admin.
Removed '.can?' method calls for checking edit & destroy
permissions on parameters and pre-calculated these permissions.

### CODE REVIEW

**Comment by kgaikwad** [comment]
@ShimShtein, 
I have pushed first version of parameter changes where I cached the `user.permissions` in `.can?` method and modified the code bit as we discussed.
Please have a look.
If there is scope of improvement, please feel free to add comments :-)

File: app/services/authorizer.rb
Comment (ares) [inline app/services/authorizer.rb]: Does this mean that user needs to relogin after admin changes his/her permissions to drop the cache?

**Review by ares — DISMISSED** [review]
IIUC the suggested change, I'm concerned with the fact we cache permission per session. That wouldn't be huge problem in case we use authorizer everywhere, but it's not the case. There some permission checks would use cache and some not. That would lead to very confusing behavior. Could the caching be added only to host params instead of modifying general authorizer code? Or is there other alternative? Or am I misreading the code? :-)

File: app/helpers/common_parameters_helper.rb
Comment (ShimShtein) [inline app/helpers/common_parameters_helper.rb]: I think we can do the same thing we do with `:view_params`:
``` ruby
{
ids_to_edit: parameters_by_type.authorized(:edit_params).pluck(:id)
ids_to_destroy: parameters_by_type.authorized(:destroy_params).pluck(:id)
}
```

File: app/helpers/common_parameters_helper.rb
Comment (ShimShtein) [inline app/helpers/common_parameters_helper.rb]: Instead of `#map` that works on returned collection of `Parameter` objects, it would be more efficient to use `#pluck`. It would generate a query that returns only the selected fields from the DB as an array.

File: app/helpers/common_parameters_helper.rb
Comment (ShimShtein) [inline app/helpers/common_parameters_helper.rb]: If we use `#authorized` scope, it wil automatically create `1=0` or `1=1` queries for these special cases. We shouldn't worry about it here.

**Review by ShimShtein — CHANGES_REQUESTED** [review]
It's a good start, but if we use `#authorized` for every action, I think we won't need changes to `autorizer` at all. It will cause 3 queries to DB (which is acceptable as it's bound).

**Comment by kgaikwad** [comment]
@ares, @ShimShtein,

Thank you for review.

> Does this mean that user needs to relogin after admin changes his/her permissions to drop the cache?


@ares, I agree with your point that there are places where it won't use cache result. so it may cause inconsistency.
Secondly, on changes into permissions of any role, I will need to touch all users object to get changed `updated_at` value. If I skip `touch` for user objects on this change, `user_id & updated_at` are not enough, so added `last_login_on` in cache_key. Now, realized that it will mislead
So I will remove this part and modify only for host_params i.e for `authorizer.can?(:create_params)`



> It's a good start, but if we use #authorized for every action, I think we won't need changes to authorizer at all. It will cause 3 queries to DB (which is acceptable as it's bound).

Yes, it will cause 3 queries to DB. So, every time we visit this page, it fires 3 queries.
Initially, I have used in a same way like `:view_params` but using `authorizer`  we can use existing cache logic to fetch result. So, later I modified it to `authorizer`. 
@ShimShtein, sorry that I am asking it again but please confirm do you want me to change it to ` #authorized`?

**Comment by kgaikwad** [comment]
@ares, @ShimShtein,
Updated  PR with suggested changes.

@ShimShtein, I am not really sure whether we can merge this logic in one [query](https://github.com/theforeman/foreman/pull/5994/files#diff-d7466ddd9140741130fc23515bce126cR93) as method [find_collection](https://github.com/theforeman/foreman/blob/develop/app/services/authorizer.rb#L27) uses multiple scopes to get that result. I am checking on it. If any pointers to improve, please let me know.

File: app/helpers/common_parameters_helper.rb
Comment (ares) [inline app/helpers/common_parameters_helper.rb]: I'm probably missins something and I'm reading this on phone but IIRC, authorizer exists exactly for this. It has separate caches for separate permissions. When it's instantiated, it can get preloaded set of objects. So I don't understand why we use it for create but bot for edit/destroy permissions here. If the collection of objects is different, we can have 3 authorizer instances.

Comment (ShimShtein) [reply app/helpers/common_parameters_helper.rb]: @ares, IIRC where this issue started from, the problem was that the cache is not filtered at all - you were getting all parameters from the DB, so it was too expensive to build that cache.

**Comment by ShimShtein** [comment]
I didn't test it yet, but from code perspective this looks good.

**Review by ShimShtein — DISMISSED** [review]

File: app/helpers/common_parameters_helper.rb
Comment (kgaikwad) [reply app/helpers/common_parameters_helper.rb]: @ShimShtein,

There are couple of changes suggested by @ares.
Basically, it includes instead of all parameters from the DB, it will limit to host's parameters.

**Comment by kgaikwad** [comment]
@ShimShtein, @ares, 

I have added changes by passing collection of specific host_parameters while creation of authorizer object.
Please let me know these  changes are valid. Especially [lines](https://github.com/theforeman/foreman/pull/5994/files#diff-2a8c1ced7ef4dfe08f35154c5a3ef0f1R28) that I have added in `find_collection` method.

File: app/helpers/common_parameters_helper.rb
Comment (ShimShtein) [inline app/helpers/common_parameters_helper.rb]: I think that `parameters_by_type` should be replaced by parameter type here. `find_collection` needs a resource type as the first parameter, not a collection.

File: app/services/authorizer.rb
Comment (ShimShtein) [inline app/services/authorizer.rb]: I think `find_collection` already knows how to short circuit these special cases in `build_filtered_scope_components`

**Review by ShimShtein — CHANGES_REQUESTED** [review]
This is a reasonable solution, I quite like it. It will need a couple of performance tests before we can merge it though.

File: app/services/authorizer.rb
Comment (kgaikwad) [reply app/services/authorizer.rb]: I think, likewise in [authorizable.rb](https://github.com/theforeman/foreman/blob/develop/app/models/concerns/authorizable.rb#L63), I need to move these two lines in helper file.

Comment (ares) [reply app/services/authorizer.rb]: `find_collection` would only be called from `can?` method, which already has a [guard for this case](https://github.com/theforeman/foreman/blob/ef4b3b25f8874b7414616acac473472fc22d0d61/app/services/authorizer.rb#L13-L15)

edit: oh sorry I missed you're calling find_collection directly

Comment (ShimShtein) [reply app/services/authorizer.rb]: @ares, i this case, should we leave the short circuit, or we can rely on `build_filtered_scope_components` to do the right job?

Comment (ares) [reply app/services/authorizer.rb]: if Authorizer#find_collection will be used, I think I'd add it explicitly here, but after yesterday chat with @kgaikwad, maybe this won't be necessary

**Comment by ShimShtein** [comment]
@kgaikwad, what's the status here?

**Comment by kgaikwad** [comment]
@ShimShtein, @ares,
Updated PR with changes where user check extracted out in separate method before calling `find_collection` in helper file.

After discussion with @ares on IRC, he suggested alternative way which I think same as what I did using `.authorized` and calculating ids per action i.e edit, destroy, view.
I am going to test both solutions against parameter data. After that, it will be easy to get which solution will be best suited here.

**Comment by kgaikwad** [comment]
@ares, @ShimShtein,
For these changes, I did some performance test with admin/non-admin user. Host edit action called twice so giving two time values for rendering partial. Here is a short summary:

**1. With current changes from develop branch.**
For admin user,
Rendered common_parameters/_parameters.html.erb (30.7ms)
Rendered common_parameters/_parameters.html.erb (28.2ms)
For non-admin user,
 Rendered common_parameters/_parameters.html.erb (32582.8ms)
 Rendered common_parameters/_parameters.html.erb (34845.5ms)

**2. With latest changes in this [pull-request](https://github.com/theforeman/foreman/pull/5994/commits/17fd8ff422d44b2f60d35ff72286c20a542a6149)**
For admin user,
Rendered common_parameters/_parameters.html.erb (32.0ms)
Rendered common_parameters/_parameters.html.erb (29.1ms)

For non-admin user, 
run 1st:
Rendered common_parameters/_parameters.html.erb (92.6ms)
Rendered common_parameters/_parameters.html.erb (43.2ms)
run 2nd:
Rendered common_parameters/_parameters.html.erb (48.8ms)
Rendered common_parameters/_parameters.html.erb (45.6ms)

**3. With changes using 3 authorizer objects in [first commit of PR](https://github.com/theforeman/foreman/pull/5994/commits/13983dd7e42f5dad506e2dcb4815d6e390595f65)**

For admin user, 
Rendered common_parameters/_parameters.html.erb (33.3ms)
Rendered common_parameters/_parameters.html.erb (34.8ms)

For non-admin user,
run 1st:
Rendered common_parameters/_parameters.html.erb (44.5ms)
Rendered common_parameters/_parameters.html.erb (39.1ms)
run 2nd:
Rendered common_parameters/_parameters.html.erb (43.8ms)
Rendered common_parameters/_parameters.html.erb (65.5ms)

Please let me know your suggestions.

Updated: From above results, result from point (3) i.e. initial commit of this PR using 3 authorizer objects gives better result than remaining two.

**Comment by ares** [comment]
Great work @kgaikwad, I think both options of the fix are comparable resource-wise. I'd prefer the later appproach (current version of PR) as it's using Authorizer code and does not introduce custom methods.

How many parameters you have defined for these benchmarks? FTR we already have integration tests (parameter_permissions.rb) for this, so probably no need to cover anything else.

**Comment by kgaikwad** [comment]
> Great work @kgaikwad, I think both options of the fix are comparable resource-wise. I'd prefer the later appproach (current version of PR) as it's using Authorizer code and does not introduce custom methods.
> 
> How many parameters you have defined for these benchmarks? FTR we already have integration tests (parameter_permissions.rb) for this, so probably no need to cover anything else.

I have used `28 host_parameters` for one host machine.

File: app/helpers/common_parameters_helper.rb
Comment (justahero) [inline app/helpers/common_parameters_helper.rb]: I would recommend to return `parameters_by_type.none` instead. It is an `ActiveRecord::Relation` acting as a Null Object that is still chainable, but does not generate a database query.

**Comment by kgaikwad** [comment]
@justahero,
Thank you for review. Added change using `.none` and fixed tests.
Tests was failing previously as current parameter objects with validation failure replaced by result parameter collection of Authorizer instead of comparing existing parameter objects with new list for view permission.

**Comment by kgaikwad** [comment]
@ShimShtein,

I have updated PR with changes we discussed about authorizer declaration and deprecation warning message.

File: app/helpers/common_parameters_helper.rb
Comment (ShimShtein) [inline app/helpers/common_parameters_helper.rb]: We are working on 1.21, so it will be deprecated at 1.23.

File: app/helpers/common_parameters_helper.rb
Comment (ShimShtein) [inline app/helpers/common_parameters_helper.rb]: Since we don't want to break existing flow for the users, we should try and "guess" the new parameters from the old ones:

``` ruby
    unless params_authorizer.is_a?(Authorizer)
      Foreman::Deprecation.deprecation_warning(
          '1.23',
          "definition changed for method 'authorized_resource_parameters'."\
          'New definition accepts Authorizer object & parameters collection as arguments.'\
          'resource and type arguments are deprecated.'\
          'Please modify method calling as per new definition.'
      )
      # Assign old parameter names and gues new parameters.
      resource = params_authorizer
      type = parameters_by_type
      parameters_by_type = resource.send(type)
      params_authorizer = Authorizer.new(User.current, :collection => parameters_by_type)
    end
```

File: app/helpers/common_parameters_helper.rb
Comment (ShimShtein) [inline app/helpers/common_parameters_helper.rb]: you don't have to use safe call here. `user` is never `nil`.

File: app/views/common_parameters/_parameters.html.erb
Comment (ShimShtein) [inline app/views/common_parameters/_parameters.html.erb]: You can create a local variable, and not a class one here, and then pass it to inner template as local:

```suggestion
        params_authorizer = Authorizer.new(User.current, :collection => parameters_by_type)
```

File: app/views/common_parameters/_parameters.html.erb
Comment (ShimShtein) [inline app/views/common_parameters/_parameters.html.erb]: Pass the authorizer as a local variable:

```suggestion
          <%= render "common_parameters/parameter", :f => builder, :authorizer => params_authorizer %>
```

File: app/views/common_parameters/_parameter.html.erb
Comment (ShimShtein) [inline app/views/common_parameters/_parameter.html.erb]: if we pass authorizer as local variable in `_parameters.html.erb`, we can use it here then:

```suggestion
  <% authorized = f.object.new_record? ? can_create_parameter?(f.object, authorizer) : authorizer.can?(:edit_params, f.object) %>
```

File: app/views/common_parameters/_parameter.html.erb
Comment (ShimShtein) [inline app/views/common_parameters/_parameter.html.erb]: same goes here:

```suggestion
    <%= link_to_remove_fields(_('Remove'), f) if f.object.new_record? || authorizer.can?(:destroy_params, f.object) %>
```

**Review by ShimShtein — CHANGES_REQUESTED** [review]
Now it looks pretty solid. A couple of small changes and it's done.

**Comment by kgaikwad** [comment]
@ShimShtein, thank you for review! 
Updated PR as per your suggestions.

**Review by ShimShtein — DISMISSED** [review]
Looks good from my point of view.
@ares, anything to add here?

**Review by ares — COMMENTED** [review]
I haven't tested but looks reasonable, @ShimShtein if you did the testing, please go ahead and merge.

File: app/helpers/common_parameters_helper.rb
Comment (tbrisker) [inline app/helpers/common_parameters_helper.rb]: we don't actually need the `parameter` parameter here

**Review by tbrisker — CHANGES_REQUESTED** [review]
Just a tiny comment inline, other that that looks like quite a great improvement!

**Comment by kgaikwad** [comment]
@tbrisker,
Good catch :+1: 
Made a suggested change to PR.

EDIT-
It is showing me that "kgaikwad dismissed ShimShtein’s stale review via afb9ace 7 minutes ago" 
is it due to force-pushed?

**Comment by tbrisker** [comment]
@kgaikwad yes, we have the repo set up so new pushes dismiss old approvals to make sure the approval is still valid. Thanks for updating this!

**Review by tbrisker — APPROVED** [review]
Thanks @kgaikwad !

**Pull request merged**
